concurrency/PubSubMessageService.java

```diff
@@ -0,0 +1,389 @@
+
+    private boolean centralExecutor = false;
+
+    //This is to avoid concurrent issue for subscribe and unsubscribe
+    private Object subscribeLock = new Object();
+
+    private static void init() {
+        try {
+            Class<?> clazz = Class.forName(subscriber_finder_class);
+
+            finder = (ISubscriberFinder) clazz.newInstance();
+            Log.info(TAG, "PubSub scanned all the subscribers");
+        } catch (Exception ex) {
+            Log.error(TAG, subscriber_finder_class + " not found");
+        }
+    }
+
+    /**
+     * get an instance of {@link IPubSubEventsManager}
+     *
+     * @return
+     */
+    public static IPubSubEventsManager getInstance() {
+        if (instance == null) {
+            synchronized (IPubSubEventsManager.class) {
+                if (instance == null) {
+                    init();
+                    instance = new PubSubMessageService();
+                }
+            }
+        }
+        return instance;
+    }
+
+    /**
+     * testing only: create a {@link IPubSubEventsManager} implementation using
+     * a custom {@link ISubscriberFinder} implementation
+     *
+     * @param finderClassName
+     * @return
+     */
+    static synchronized IPubSubEventsManager getInstance(String finderClassName) {
+        subscriber_finder_class = finderClassName;
+        instance = null;
+        return getInstance();
+    }
+
+    PubSubMessageService() {
+        this.centralExecutor = BuildInfo.isEInkBuild();
+        if (this.centralExecutor) {
+            eventExecutor = newExecutor();
+        }
+    }
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see
+     * com.amazon.kindle.krx.events.IPubSubEventsManager#createMessageQueue(
+     * java.lang.Class)
+     */
+    @Override
+    public IMessageQueue createMessageQueue(Class<?> publisher) {
+        if (centralExecutor) {
+            return messageQueue;
+        }
+        return createMessageQueue(publisher, null);
+    }
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see
+     * com.amazon.kindle.krx.events.IPubSubEventsManager#createMessageQueue(
+     * java.lang.Class, java.util.concurrent.ExecutorService)
+     */
+    @Override
+    @VisibleForTesting
+    public synchronized IMessageQueue createMessageQueue(Class<?> publisher,
+                                                         ExecutorService executor) {
+        IMessageQueue queue = messageQueues.get(publisher);
+        if (queue == null) {
+            queue = new MessageQueue(this, executor,
+                    deadMessageHandler);
+            messageQueues.put(publisher, queue);
+        }
+
+        return queue;
+    }
+
+    @Override
+    public ITopicMessageQueue createTopicMessageQueue(String topic) {
+        ITopicMessageQueue queue = topicToQueue.get(topic);
+        if (queue == null) {
+            queue = new TopicMessageQueue(topic, this, null, deadMessageHandler);
+            topicToQueue.put(topic, queue);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **166, 167, and 169**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **172**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.